src/library/scala/collection/convert/JavaCollectionWrappers.scala

```diff
@@ -462,8 +480,17 @@ private[collection] object JavaCollectionWrappers extends Serializable {
         case _ => Try(last).toOption
       }
 
-    override def updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = Option {
-      underlying.compute(key, (_, v) => remappingFunction(Option(v)).getOrElse(null.asInstanceOf[V]))
+    override def updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = {
+      def remap(k: K, v: V): V =
+        remappingFunction(Option(v)) match {
+          case Some(null) => throw PutNull
+          case Some(x)    => x
+          case None       => null.asInstanceOf[V]
+        }
+      try Option(underlying.compute(key, remap))
+      catch {
+        case PutNull => super.updateWith(key)(remappingFunction)
```

Comment:
So the dance is between
  - supporting Scala semantics for `updateWith` (can put `Some(null)`)
  - supporting the underlying map's contract (the `ConcurrentHashMap.compute` method evaluates the argument function only once)

and we can't have both, so for `Some(null)` the function will re-evaluate. That looks fine to me.